# Домашняя работа по теме "Машинное обучение ранжированию"

В этом ДЗ мы:
- научимся работать со стандартным датасетом для машинного обучения ранжированию [MSLR](https://www.microsoft.com/en-us/research/project/mslr/)
- попробуем применить на практике все то, чему мы научились на семинаре

## Как будет происходить сдача ДЗ

Вам надо:
- форкнуть эту репу
- создать бранч в котором вы дальше будете работать
- реализовать класс Model в этом ноутбуке
- убедиться, что ваша реализация выбивает NDCG@10 выше бейзлайна (см. ниже)
- запушить ваш бранч и поставить Pull Request
- в комментарии написать какой скор вы выбили

В таком случае мы (организаторы):

- счекаутим вашу бранчу
- проверим что ваша реализация действительно выбивает заявленный скор

Предполагается, что и вы, и мы работаем в виртаульном окружении как в семинаре про машинное обучение ранжированию: seminars/7-learning-to-rank/requirements.txt(подробнее про работу с виртуальными окружениями README в корне этой репы).

Оценка:
- За выбитый скор больше **0.507** назначаем **5** баллов, за скор больше (или равно) **0.510** назначаем максимальный балл -- 10 баллов
- Тот из участников кто выбъет самый высокий скор получит еще +10 баллов

При сдаче кода важно помнить о том, что:
- В коде не должно быть захардкоженных с потолка взятых гиперпараметров (таких как число деревьев, learning rate и т.п.) -- обязательно должен быть представлен код который их подбирает!
- Решение должно быть стабильно от запуска к запуску (на CPU) т.е. все seed'ы для генераторов случайных чисел должны быть фиксированы
- Мы (организаторы) будем запускать код на CPU поэтому, даже если вы использовали для подбора параметров GPU, финальный скор надо репортить на CPU

## Пререквизиты

Импортируем все что нам понадобится для дальнейшей работы:

In [1]:
import pathlib
from timeit import default_timer as timer

import numpy as np
import pandas as pd
from tqdm import tqdm
import catboost
import copy
import random
import pickle

from catboost import datasets, utils

## Датасет MSLR (Microsoft Learning to Rank)

Загрузим датасет MSLR.

Полный датасет можно скачать с официального сайта: https://www.microsoft.com/en-us/research/project/mslr/

Строго говоря, он состоит их 2х частей:

- основной датасет MSLR-WEB30K -- он содержит более 30 тыс. запросов
- "маленький" датасет MSLR-WEB10K, который содержит только 10 тыс. запросов и является случайным сэмплом датасета MSLR-WEB30K

в этом ДЗ мы будем работать с MSLR-WEB10K, т.к. полная версия датасета может просто не поместиться у нас в RAM (и, тем более, в память видеокарты если мы учимся на GPU)

Будем считать, что мы самостоятельно скачали датасет MSLR-WEB10K с официального сайта, поместили его в папку КОРЕНЬ-ЭТОЙ-РЕПЫ/data/mslr-web10k и раззиповали.

В результате у нас должна получиться следующая структура папок:

In [2]:
!ls /kaggle/input/vk-hw3-ranking/MSLR-WEB10K

Fold1


Заметим, что датасет довольно большой, в распакованном виде он весит 7.7 GB.

Датасет состоит из нескольких фолдов, которые по сути представляют из себя разные разбиения одних и тех же данных на обучающее, валидационное и тестовые множеста.

Дальше мы будем использовать только первый фолд: Fold1.

Заглянем внутрь:

In [3]:
!ls -lh /kaggle/input/vk-hw3-ranking/MSLR-WEB10K/Fold1

total 1.3G
-rw-r--r-- 1 nobody nogroup 267M Apr 13 19:15 test.txt
-rw-r--r-- 1 nobody nogroup 800M Apr 13 19:16 train.txt
-rw-r--r-- 1 nobody nogroup 261M Apr 13 19:15 vali.txt


Видим, что у нас 3 файла с говорящими названиями, соответсвующими сплитам нашего датасета.

Посмотрим на содержимое одного из файлов:

In [4]:
!head -n 1 /kaggle/input/vk-hw3-ranking/MSLR-WEB10K/Fold1/train.txt

2 qid:1 1:3 2:3 3:0 4:0 5:3 6:1 7:1 8:0 9:0 10:1 11:156 12:4 13:0 14:7 15:167 16:6.931275 17:22.076928 18:19.673353 19:22.255383 20:6.926551 21:3 22:3 23:0 24:0 25:6 26:1 27:1 28:0 29:0 30:2 31:1 32:1 33:0 34:0 35:2 36:1 37:1 38:0 39:0 40:2 41:0 42:0 43:0 44:0 45:0 46:0.019231 47:0.75000 48:0 49:0 50:0.035928 51:0.00641 52:0.25000 53:0 54:0 55:0.011976 56:0.00641 57:0.25000 58:0 59:0 60:0.011976 61:0.00641 62:0.25000 63:0 64:0 65:0.011976 66:0 67:0 68:0 69:0 70:0 71:6.931275 72:22.076928 73:0 74:0 75:13.853103 76:1.152128 77:5.99246 78:0 79:0 80:2.297197 81:3.078917 82:8.517343 83:0 84:0 85:6.156595 86:2.310425 87:7.358976 88:0 89:0 90:4.617701 91:0.694726 92:1.084169 93:0 94:0 95:2.78795 96:1 97:1 98:0 99:0 100:1 101:1 102:1 103:0 104:0 105:1 106:12.941469 107:20.59276 108:0 109:0 110:16.766961 111:-18.567793 112:-7.760072 113:-20.838749 114:-25.436074 115:-14.518523 116:-21.710022 117:-21.339609 118:-24.497864 119:-27.690319 120:-20.203779 121:-15.449379 122:-4.474452 123:-23.634899 

Видим, что данные лежат в уже знакомом нам по семинару формате:

- В первой колонке лежит таргет (оценка асессора), по 5-балльной шкале релевантности: от 0 до 4 (включительно)
- Во второй колонке лежит ID запроса, по которому можно сгруппировать все оценки документов в рамках одного и того же запроса
- Дальше идет вектор из 128 фичей (таких как значения BM25 и т.п.), их точная природа нам сейчас на важна

В файле qid и все-фичи кодируются в формате КЛЮЧ:ЗНАЧЕНИЕ, напр. 130:116 -- тут 130 это номер фичи, а 116 -- ее значение.

Такой формат в мире машинного обучения часто называют svm light формат (в честь когда-то популярной библиотеки SVM-Light)

Напишем немного вспомогательного кода для загрузки этого датасета:

In [5]:
def generate_column_names(num_features):
    """Generates column names for LETOR-like datasets"""
    columns = ['label', 'qid']
    for i in range(num_features):
        column = f"feature_{i+1}"
        columns.append(column)
    return columns
    
def load_svmlight_file(input_file, max_num_lines=0):
    """Loads dataset split in SVM-Light format"""
    def _parse_field(field):
        parts = field.split(':')
        if len(parts) != 2:
            raise Exception(f"invalid number of parts in field {field}")
        return parts

    num_features = 136
    exp_num_fields = num_features + 2
    num_lines = 0
    X = []
    with open(input_file, 'rt') as f:
        for line in f:
            try:
                num_lines += 1
                                  
                # Parse into fields
                fields = line.rstrip().split(' ')
                num_fields = len(fields)
                if num_fields != exp_num_fields:
                    raise Exception(f"invalid number of fields {num_fields}")
    
                # Parse every field
                x = np.zeros(exp_num_fields, dtype=np.float32)
                label = int(fields[0])
                x[0] = label
                _, qid_str = _parse_field(fields[1])
                qid = int(qid_str)
                x[1] = qid
                for i, field in enumerate(fields[2:]):
                    _, feature_str = _parse_field(field)
                    x[i+2] = float(feature_str)
    
                # Add new object
                X.append(x)
                if num_lines % 50000 == 0:
                    print(f"Loaded {num_lines} lines...")
                if max_num_lines > 0 and num_lines == max_num_lines:
                    print(f"WARNING: stop loading, line limit reached: max_num_lines = {max_num_lines} input_file = {input_file}")
                    break
            except Exception as e:
                raise Exception(f"error at line {num_lines} in {input_file}") from e
    
    # To pandas
    print(f'Creating pandas dataframe with size = {len(X)}')
    df = pd.DataFrame(X, columns=generate_column_names(num_features))
    print(f"Loaded SVM-Light file {input_file}")
    return df

И теперь загрузим датасет:

In [6]:
fold_dir = pathlib.Path("/kaggle/input/vk-hw3-ranking/MSLR-WEB10K/Fold1")

df_train = load_svmlight_file(fold_dir.joinpath("train.txt"))
df_valid = load_svmlight_file(fold_dir.joinpath("vali.txt"))
df_test = load_svmlight_file(fold_dir.joinpath("test.txt"))
print(f"Dataset loaded from fold_dir {fold_dir}")

Loaded 50000 lines...
Loaded 100000 lines...
Loaded 150000 lines...
Loaded 200000 lines...
Loaded 250000 lines...
Loaded 300000 lines...
Loaded 350000 lines...
Loaded 400000 lines...
Loaded 450000 lines...
Loaded 500000 lines...
Loaded 550000 lines...
Loaded 600000 lines...
Loaded 650000 lines...
Loaded 700000 lines...
Creating pandas dataframe with size = 723412
Loaded SVM-Light file /kaggle/input/vk-hw3-ranking/MSLR-WEB10K/Fold1/train.txt
Loaded 50000 lines...
Loaded 100000 lines...
Loaded 150000 lines...
Loaded 200000 lines...
Creating pandas dataframe with size = 235259
Loaded SVM-Light file /kaggle/input/vk-hw3-ranking/MSLR-WEB10K/Fold1/vali.txt
Loaded 50000 lines...
Loaded 100000 lines...
Loaded 150000 lines...
Loaded 200000 lines...
Creating pandas dataframe with size = 241521
Loaded SVM-Light file /kaggle/input/vk-hw3-ranking/MSLR-WEB10K/Fold1/test.txt
Dataset loaded from fold_dir /kaggle/input/vk-hw3-ranking/MSLR-WEB10K/Fold1


Посмотрим на данные:

In [7]:
df_train.head(5)

label  qid  feature_1  feature_2  feature_3  feature_4  feature_5  \
0    2.0  1.0        3.0        3.0        0.0        0.0        3.0   
1    2.0  1.0        3.0        0.0        3.0        0.0        3.0   
2    0.0  1.0        3.0        0.0        2.0        0.0        3.0   
3    2.0  1.0        3.0        0.0        3.0        0.0        3.0   
4    1.0  1.0        3.0        0.0        3.0        0.0        3.0   

   feature_6  feature_7  feature_8  ...  feature_127  feature_128  \
0        1.0        1.0   0.000000  ...         62.0   11089534.0   
1        1.0        0.0   1.000000  ...         54.0   11089534.0   
2        1.0        0.0   0.666667  ...         45.0          3.0   
3        1.0        0.0   1.000000  ...         56.0   11089534.0   
4        1.0        0.0   1.000000  ...         64.0          5.0   

   feature_129  feature_130  feature_131  feature_132  feature_133  \
0          2.0        116.0      64034.0         13.0          3.0   
1          2.0        124.0      64034.0          1.0          2.0   
2          1.0        124.0       3344.0         14.0         67.0   
3         13.0        123.0      63933.0          1.0          3.0   
4          7.0        256.0      49697.0          1.0         13.0   

   feature_134  feature_135  feature_136  
0          0.0          0.0          0.0  
1          0.0          0.0          0.0  
2          0.0          0.0          0.0  
3          0.0          0.0          0.0  
4          0.0          0.0          0.0  

[5 rows x 138 columns]

Т.е. теперь мы видим что данные доступны в точно таком же виде, как это было в семинаре.

Проведем небольшой EDA.

Всего у нас 723 тыс. документов в трейне:

In [8]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 723412 entries, 0 to 723411
Columns: 138 entries, label to feature_136
dtypes: float32(138)
memory usage: 380.8 MB


235 тыс. документов в валидации:

In [9]:
df_valid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235259 entries, 0 to 235258
Columns: 138 entries, label to feature_136
dtypes: float32(138)
memory usage: 123.8 MB


И 241 тыс. документов в тесте:

In [10]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 241521 entries, 0 to 241520
Columns: 138 entries, label to feature_136
dtypes: float32(138)
memory usage: 127.1 MB


Сколько у нас запросов?

In [11]:
num_queries_train = df_train['qid'].nunique()
num_queries_valid = df_valid['qid'].nunique()
num_queries_test = df_test['qid'].nunique()
print(f"Got {num_queries_train} train, {num_queries_valid} valid and {num_queries_test} test queries")

Got 6000 train, 2000 valid and 2000 test queries


## Обучаем модель

In [12]:
def to_catboost_dataset(df):
    y = df['label'].to_numpy()                       # Label: [0-4]
    q = df['qid'].to_numpy().astype('uint32')        # Query Id
    X = df.drop(columns=['label', 'qid']).to_numpy() # 136 features
    return (X, y, q)

X_train, y_train, q_train = to_catboost_dataset(df_train)
X_valid, y_valid, q_valid = to_catboost_dataset(df_valid)
X_test, y_test, q_test = to_catboost_dataset(df_test)

In [51]:
pool_train = catboost.Pool(data=X_train, label=y_train, group_id=q_train)
pool_valid = catboost.Pool(data=X_valid, label=y_valid, group_id=q_valid)

In [56]:
def search(n_iter, pool_train, pool_valid):
    np.random.seed(0)
    random.seed(0)
    best_score = 0
    best_model = None
    best_params = {}
    y_valid, q_valid = pool_valid.get_label(), pool_valid.get_group_id_hash()
    
    for i in range(n_iter):
        print(f'{"-"*20} Model number {i+1} {"-"*20}')
        params = {
                'learning_rate': np.random.uniform(1e-3, 1e-1),
                'depth': random.randint(3, 8),
                'l2_leaf_reg': np.random.uniform(1, 30),
                'loss_function': 'YetiRank',
                'eval_metric': 'NDCG:top=10;type=Exp',
                'iterations': 1000,
                'bagging_temperature': np.random.uniform(0.7, 1),
                'use_best_model': True,
                'min_data_in_leaf': random.randint(1, 100),
                'verbose': 200,
                'grow_policy': random.choice(['SymmetricTree', 'Depthwise', 'Lossguide']),
                #'max_leaves': random.randint(10, 50),
                'task_type': "GPU",
               }
        print('Params:')
        for k,v in params.items():
            print(f'\t - {k}: {v}')
        model = catboost.CatBoost(params)
        start = timer()
        model.fit(pool_train, eval_set=pool_valid)
        score = utils.eval_metric(y_valid, model.predict(pool_valid), 
                                  'NDCG:top=10;type=Exp', group_id=q_valid)[0]
        print(f'My score = {score}')
        if score > best_score:
            best_model = model
            best_score = score
            best_params = params
            
        elapsed = timer() - start
        print(f"Model fit: elapsed = {elapsed:.3f} with test/NDCG@10 = {score}")
        
    return best_model, best_params

In [57]:
best_model, best_params = search(25, pool_train, pool_valid)

-------------------- Model number 1 --------------------
Params:
	 - learning_rate: 0.055332536888805156
	 - depth: 6
	 - l2_leaf_reg: 21.740491624800164
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8808290128214932
	 - use_best_model: True
	 - min_data_in_leaf: 98
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3312943	best: 0.3312943 (0)	total: 51.9ms	remaining: 51.8s
200:	test: 0.4803541	best: 0.4803541 (200)	total: 6.49s	remaining: 25.8s
400:	test: 0.4947174	best: 0.4947174 (400)	total: 12.8s	remaining: 19.1s
600:	test: 0.5009019	best: 0.5009019 (600)	total: 19.1s	remaining: 12.7s
800:	test: 0.5041915	best: 0.5041915 (800)	total: 25.4s	remaining: 6.31s
999:	test: 0.5064371	best: 0.5066606 (990)	total: 31.5s	remaining: 0us
bestTest = 0.506660638
bestIteration = 990
Shrink model to first 991 iterations.
My score = 0.5066606379658462
Model fit: elapsed = 36.291 with test/NDCG@10 = 0.5066606379658462
-------------------- Model number 2 --------------------
Params:
	 - learning_rate: 0.05494343511669279
	 - depth: 3
	 - l2_leaf_reg: 13.285989180828237
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8937682339199968
	 - use_best_model: True
	 - min_data_in_leaf: 34
	 - verbose: 200
	 - grow_policy: Lossguide
	 - task_ty

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2474886	best: 0.2474886 (0)	total: 42.4ms	remaining: 42.3s
200:	test: 0.4533229	best: 0.4533229 (200)	total: 5.23s	remaining: 20.8s
400:	test: 0.4703910	best: 0.4703910 (400)	total: 10.4s	remaining: 15.6s
600:	test: 0.4781788	best: 0.4781788 (600)	total: 15.6s	remaining: 10.4s
800:	test: 0.4836509	best: 0.4836509 (800)	total: 20.8s	remaining: 5.18s
999:	test: 0.4864931	best: 0.4866632 (995)	total: 25.9s	remaining: 0us
bestTest = 0.4866631814
bestIteration = 995
Shrink model to first 996 iterations.
My score = 0.48666318135940495
Model fit: elapsed = 29.540 with test/NDCG@10 = 0.48666318135940495
-------------------- Model number 3 --------------------
Params:
	 - learning_rate: 0.04432113391500656
	 - depth: 6
	 - l2_leaf_reg: 26.861417022680314
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.9890988281503088
	 - use_best_model: True
	 - min_data_in_leaf: 52
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3255623	best: 0.3255623 (0)	total: 49.4ms	remaining: 49.3s
200:	test: 0.4751316	best: 0.4751316 (200)	total: 6.49s	remaining: 25.8s
400:	test: 0.4894836	best: 0.4894836 (400)	total: 12.9s	remaining: 19.3s
600:	test: 0.4966342	best: 0.4967753 (595)	total: 19.2s	remaining: 12.7s
800:	test: 0.5011411	best: 0.5011411 (800)	total: 25.5s	remaining: 6.33s
999:	test: 0.5035916	best: 0.5035979 (985)	total: 31.7s	remaining: 0us
bestTest = 0.503597886
bestIteration = 985
Shrink model to first 986 iterations.
My score = 0.5035978859742678
Model fit: elapsed = 36.506 with test/NDCG@10 = 0.5035978859742678
-------------------- Model number 4 --------------------
Params:
	 - learning_rate: 0.038960710363751996
	 - depth: 6
	 - l2_leaf_reg: 23.960026104397272
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8586684759258714
	 - use_best_model: True
	 - min_data_in_leaf: 46
	 - verbose: 200
	 - grow_policy: Lossguide
	 - task_t

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2871093	best: 0.2871093 (0)	total: 51.5ms	remaining: 51.4s
200:	test: 0.4685307	best: 0.4685307 (200)	total: 6.69s	remaining: 26.6s
400:	test: 0.4862028	best: 0.4862028 (400)	total: 13.3s	remaining: 19.9s
600:	test: 0.4949805	best: 0.4949805 (600)	total: 20.1s	remaining: 13.3s
800:	test: 0.4986694	best: 0.4987613 (775)	total: 26.6s	remaining: 6.61s
999:	test: 0.5024616	best: 0.5025868 (995)	total: 33.1s	remaining: 0us
bestTest = 0.5025867875
bestIteration = 995
Shrink model to first 996 iterations.
My score = 0.5025867874532144
Model fit: elapsed = 37.668 with test/NDCG@10 = 0.5025867874532144
-------------------- Model number 5 --------------------
Params:
	 - learning_rate: 0.0572364115482993
	 - depth: 4
	 - l2_leaf_reg: 27.84230251048717
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.7213108174593661
	 - use_best_model: True
	 - min_data_in_leaf: 65
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2507868	best: 0.2507868 (0)	total: 8.01s	remaining: 2h 13m 26s
200:	test: 0.4475551	best: 0.4475551 (200)	total: 13.5s	remaining: 53.6s
400:	test: 0.4634765	best: 0.4634765 (400)	total: 18.8s	remaining: 28.1s
600:	test: 0.4733441	best: 0.4734197 (595)	total: 24.1s	remaining: 16s
800:	test: 0.4796564	best: 0.4797753 (795)	total: 29.5s	remaining: 7.32s
999:	test: 0.4841196	best: 0.4841196 (999)	total: 34.7s	remaining: 0us
bestTest = 0.4841196112
bestIteration = 999
My score = 0.48411961122918845
Model fit: elapsed = 37.081 with test/NDCG@10 = 0.48411961122918845
-------------------- Model number 6 --------------------
Params:
	 - learning_rate: 0.00962580067045253
	 - depth: 5
	 - l2_leaf_reg: 1.586333525769446
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.9497859536643813
	 - use_best_model: True
	 - min_data_in_leaf: 18
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2608959	best: 0.2608959 (0)	total: 42.7ms	remaining: 42.7s
200:	test: 0.4120193	best: 0.4120395 (195)	total: 5.75s	remaining: 22.9s
400:	test: 0.4280165	best: 0.4280165 (400)	total: 11.4s	remaining: 17s
600:	test: 0.4392903	best: 0.4392903 (600)	total: 17s	remaining: 11.3s
800:	test: 0.4490987	best: 0.4490987 (800)	total: 22.6s	remaining: 5.62s
999:	test: 0.4550753	best: 0.4550753 (999)	total: 28.2s	remaining: 0us
bestTest = 0.4550752807
bestIteration = 999
My score = 0.4550752807180393
Model fit: elapsed = 30.795 with test/NDCG@10 = 0.4550752807180393
-------------------- Model number 7 --------------------
Params:
	 - learning_rate: 0.0780375183440352
	 - depth: 7
	 - l2_leaf_reg: 26.230352299157754
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.9935855026698293
	 - use_best_model: True
	 - min_data_in_leaf: 33
	 - verbose: 200
	 - grow_policy: Lossguide
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2855559	best: 0.2855559 (0)	total: 51.6ms	remaining: 51.5s
200:	test: 0.4857467	best: 0.4857467 (200)	total: 7s	remaining: 27.8s
400:	test: 0.4987151	best: 0.4987151 (400)	total: 13.8s	remaining: 20.6s
600:	test: 0.5036730	best: 0.5037669 (590)	total: 20.5s	remaining: 13.6s
800:	test: 0.5067882	best: 0.5068620 (795)	total: 27.1s	remaining: 6.73s
999:	test: 0.5078572	best: 0.5081747 (995)	total: 33.7s	remaining: 0us
bestTest = 0.5081746544
bestIteration = 995
Shrink model to first 996 iterations.
My score = 0.5081746544008751
Model fit: elapsed = 38.883 with test/NDCG@10 = 0.5081746544008751
-------------------- Model number 8 --------------------
Params:
	 - learning_rate: 0.08011669785745563
	 - depth: 8
	 - l2_leaf_reg: 14.382901505335024
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.9341587528859366
	 - use_best_model: True
	 - min_data_in_leaf: 78
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task_

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3075570	best: 0.3075570 (0)	total: 53.8ms	remaining: 53.7s
200:	test: 0.4709863	best: 0.4709863 (200)	total: 6.77s	remaining: 26.9s
400:	test: 0.4881581	best: 0.4881581 (400)	total: 13.5s	remaining: 20.1s
600:	test: 0.4956842	best: 0.4956842 (600)	total: 20.2s	remaining: 13.4s
800:	test: 0.4991913	best: 0.4991913 (800)	total: 26.9s	remaining: 6.69s
999:	test: 0.5010748	best: 0.5012667 (990)	total: 33.7s	remaining: 0us
bestTest = 0.5012667403
bestIteration = 990
Shrink model to first 991 iterations.
My score = 0.5012667402768924
Model fit: elapsed = 36.440 with test/NDCG@10 = 0.5012667402768924
-------------------- Model number 9 --------------------
Params:
	 - learning_rate: 0.012709168161024388
	 - depth: 5
	 - l2_leaf_reg: 19.55770961849819
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.7430059862227139
	 - use_best_model: True
	 - min_data_in_leaf: 13
	 - verbose: 200
	 - grow_policy: Lossguide
	 - task_t

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2809693	best: 0.2809693 (0)	total: 46.4ms	remaining: 46.4s
200:	test: 0.4392211	best: 0.4392211 (200)	total: 6.48s	remaining: 25.8s
400:	test: 0.4538185	best: 0.4538185 (400)	total: 12.8s	remaining: 19.2s
600:	test: 0.4641042	best: 0.4641042 (600)	total: 19.2s	remaining: 12.7s
800:	test: 0.4714468	best: 0.4714468 (800)	total: 25.6s	remaining: 6.36s
999:	test: 0.4771956	best: 0.4771956 (999)	total: 31.9s	remaining: 0us
bestTest = 0.4771955542
bestIteration = 999
My score = 0.4771955542351021
Model fit: elapsed = 36.372 with test/NDCG@10 = 0.4771955542351021
-------------------- Model number 10 --------------------
Params:
	 - learning_rate: 0.09452222278790881
	 - depth: 3
	 - l2_leaf_reg: 16.13360133075208
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8243985819971571
	 - use_best_model: True
	 - min_data_in_leaf: 88
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2724801	best: 0.2724801 (0)	total: 42.4ms	remaining: 42.4s
200:	test: 0.4674195	best: 0.4674195 (200)	total: 5.26s	remaining: 20.9s
400:	test: 0.4801042	best: 0.4803098 (395)	total: 10.5s	remaining: 15.7s
600:	test: 0.4863022	best: 0.4865666 (580)	total: 15.6s	remaining: 10.4s
800:	test: 0.4911534	best: 0.4912283 (795)	total: 20.9s	remaining: 5.18s
999:	test: 0.4939991	best: 0.4941181 (995)	total: 26.1s	remaining: 0us
bestTest = 0.4941180606
bestIteration = 995
Shrink model to first 996 iterations.
My score = 0.4941180605723142
Model fit: elapsed = 29.816 with test/NDCG@10 = 0.4941180605723142
-------------------- Model number 11 --------------------
Params:
	 - learning_rate: 0.027191005598358072
	 - depth: 6
	 - l2_leaf_reg: 23.452776993592284
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8368450996649646
	 - use_best_model: True
	 - min_data_in_leaf: 72
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - 

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2795641	best: 0.2795641 (0)	total: 44.8ms	remaining: 44.7s
200:	test: 0.4320935	best: 0.4320935 (200)	total: 6s	remaining: 23.8s
400:	test: 0.4545652	best: 0.4545652 (400)	total: 12s	remaining: 17.9s
600:	test: 0.4648032	best: 0.4648032 (600)	total: 17.9s	remaining: 11.9s
800:	test: 0.4737907	best: 0.4737907 (800)	total: 23.9s	remaining: 5.94s
999:	test: 0.4793685	best: 0.4793685 (999)	total: 29.9s	remaining: 0us
bestTest = 0.4793685346
bestIteration = 999
My score = 0.4793685345601686
Model fit: elapsed = 32.344 with test/NDCG@10 = 0.4793685345601686
-------------------- Model number 12 --------------------
Params:
	 - learning_rate: 0.05727496093799621
	 - depth: 5
	 - l2_leaf_reg: 1.5449042126542991
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8852906491227631
	 - use_best_model: True
	 - min_data_in_leaf: 56
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3234188	best: 0.3234188 (0)	total: 45.4ms	remaining: 45.3s
200:	test: 0.4797761	best: 0.4797761 (200)	total: 5.97s	remaining: 23.7s
400:	test: 0.4934798	best: 0.4935515 (395)	total: 11.9s	remaining: 17.8s
600:	test: 0.4993690	best: 0.4994620 (595)	total: 17.8s	remaining: 11.8s
800:	test: 0.5023687	best: 0.5026618 (775)	total: 23.7s	remaining: 5.88s
999:	test: 0.5046248	best: 0.5046248 (999)	total: 29.4s	remaining: 0us
bestTest = 0.5046247679
bestIteration = 999
My score = 0.5046247678965097
Model fit: elapsed = 33.903 with test/NDCG@10 = 0.5046247678965097
-------------------- Model number 13 --------------------
Params:
	 - learning_rate: 0.06159747654951973
	 - depth: 7
	 - l2_leaf_reg: 18.891085909367952
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.9831244235543872
	 - use_best_model: True
	 - min_data_in_leaf: 82
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2921994	best: 0.2921994 (0)	total: 50.8ms	remaining: 50.7s
200:	test: 0.4609924	best: 0.4609924 (200)	total: 6.35s	remaining: 25.3s
400:	test: 0.4803829	best: 0.4803829 (400)	total: 12.6s	remaining: 18.8s
600:	test: 0.4879374	best: 0.4879374 (600)	total: 18.8s	remaining: 12.5s
800:	test: 0.4941941	best: 0.4942461 (795)	total: 25.1s	remaining: 6.24s
999:	test: 0.4971519	best: 0.4971519 (999)	total: 31.3s	remaining: 0us
bestTest = 0.4971518749
bestIteration = 999
My score = 0.49715187486874696
Model fit: elapsed = 33.824 with test/NDCG@10 = 0.49715187486874696
-------------------- Model number 14 --------------------
Params:
	 - learning_rate: 0.06850020961124487
	 - depth: 7
	 - l2_leaf_reg: 11.425729116639793
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8311095861398023
	 - use_best_model: True
	 - min_data_in_leaf: 62
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3445644	best: 0.3445644 (0)	total: 54.8ms	remaining: 54.8s
200:	test: 0.4901607	best: 0.4901607 (200)	total: 7.09s	remaining: 28.2s
400:	test: 0.5023699	best: 0.5023699 (400)	total: 14.1s	remaining: 21s
600:	test: 0.5064009	best: 0.5064009 (600)	total: 21s	remaining: 13.9s
800:	test: 0.5086292	best: 0.5086292 (800)	total: 27.7s	remaining: 6.88s
999:	test: 0.5092522	best: 0.5092522 (999)	total: 34.4s	remaining: 0us
bestTest = 0.5092522139
bestIteration = 999
My score = 0.5092522138519522
Model fit: elapsed = 39.753 with test/NDCG@10 = 0.5092522138519522
-------------------- Model number 15 --------------------
Params:
	 - learning_rate: 0.07006548839679923
	 - depth: 7
	 - l2_leaf_reg: 2.746538677248825
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.9000300146337002
	 - use_best_model: True
	 - min_data_in_leaf: 34
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3099908	best: 0.3099908 (0)	total: 50.9ms	remaining: 50.8s
200:	test: 0.4730758	best: 0.4730758 (200)	total: 6.29s	remaining: 25s
400:	test: 0.4907330	best: 0.4907330 (400)	total: 12.6s	remaining: 18.9s
600:	test: 0.4969699	best: 0.4969699 (600)	total: 18.9s	remaining: 12.5s
800:	test: 0.5001164	best: 0.5001482 (780)	total: 25.1s	remaining: 6.23s
999:	test: 0.5027448	best: 0.5027448 (999)	total: 31.3s	remaining: 0us
bestTest = 0.5027448268
bestIteration = 999
My score = 0.5027448268325874
Model fit: elapsed = 33.757 with test/NDCG@10 = 0.5027448268325874
-------------------- Model number 16 --------------------
Params:
	 - learning_rate: 0.06739314909219779
	 - depth: 7
	 - l2_leaf_reg: 7.1010942711413865
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.738677889296456
	 - use_best_model: True
	 - min_data_in_leaf: 2
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3079272	best: 0.3079272 (0)	total: 47.7ms	remaining: 47.6s
200:	test: 0.4681984	best: 0.4681984 (200)	total: 6.31s	remaining: 25.1s
400:	test: 0.4874033	best: 0.4874451 (395)	total: 12.6s	remaining: 18.8s
600:	test: 0.4938404	best: 0.4939996 (575)	total: 18.9s	remaining: 12.5s
800:	test: 0.4976784	best: 0.4977097 (795)	total: 25.1s	remaining: 6.23s
999:	test: 0.5001159	best: 0.5002953 (955)	total: 31.3s	remaining: 0us
bestTest = 0.5002953086
bestIteration = 955
Shrink model to first 956 iterations.
My score = 0.5002953086011998
Model fit: elapsed = 33.900 with test/NDCG@10 = 0.5002953086011998
-------------------- Model number 17 --------------------
Params:
	 - learning_rate: 0.032227406741494206
	 - depth: 8
	 - l2_leaf_reg: 11.547612357336055
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8710590311253639
	 - use_best_model: True
	 - min_data_in_leaf: 52
	 - verbose: 200
	 - grow_policy: Lossguide
	 - task

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2862021	best: 0.2862021 (0)	total: 55.2ms	remaining: 55.1s
200:	test: 0.4656991	best: 0.4656991 (200)	total: 7.21s	remaining: 28.7s
400:	test: 0.4838406	best: 0.4838406 (400)	total: 14.2s	remaining: 21.3s
600:	test: 0.4944676	best: 0.4944676 (600)	total: 21.2s	remaining: 14.1s
800:	test: 0.4989270	best: 0.4992236 (795)	total: 28.2s	remaining: 7s
999:	test: 0.5024772	best: 0.5024772 (999)	total: 35s	remaining: 0us
bestTest = 0.5024772395
bestIteration = 999
My score = 0.5024772394770559
Model fit: elapsed = 40.629 with test/NDCG@10 = 0.5024772394770559
-------------------- Model number 18 --------------------
Params:
	 - learning_rate: 0.04442154983276972
	 - depth: 8
	 - l2_leaf_reg: 29.66284130371756
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.7306134432244084
	 - use_best_model: True
	 - min_data_in_leaf: 81
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2738333	best: 0.2738333 (0)	total: 56.4ms	remaining: 56.3s
200:	test: 0.4473494	best: 0.4473494 (200)	total: 6.77s	remaining: 26.9s
400:	test: 0.4687501	best: 0.4687501 (400)	total: 13.5s	remaining: 20.1s
600:	test: 0.4796128	best: 0.4796128 (600)	total: 20.2s	remaining: 13.4s
800:	test: 0.4871781	best: 0.4872395 (795)	total: 26.8s	remaining: 6.67s
999:	test: 0.4914757	best: 0.4914757 (999)	total: 33.5s	remaining: 0us
bestTest = 0.4914756582
bestIteration = 999
My score = 0.49147565817445654
Model fit: elapsed = 36.000 with test/NDCG@10 = 0.49147565817445654
-------------------- Model number 19 --------------------
Params:
	 - learning_rate: 0.021678798853388637
	 - depth: 7
	 - l2_leaf_reg: 5.677976018664891
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8959324976396195
	 - use_best_model: True
	 - min_data_in_leaf: 64
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3479259	best: 0.3479259 (0)	total: 54ms	remaining: 54s
200:	test: 0.4728500	best: 0.4728500 (200)	total: 7.13s	remaining: 28.3s
400:	test: 0.4832577	best: 0.4832577 (400)	total: 14.1s	remaining: 21.1s
600:	test: 0.4915512	best: 0.4915512 (600)	total: 21.1s	remaining: 14s
800:	test: 0.4972673	best: 0.4972673 (800)	total: 28.1s	remaining: 6.99s
999:	test: 0.5000877	best: 0.5001758 (995)	total: 35s	remaining: 0us
bestTest = 0.5001758194
bestIteration = 995
Shrink model to first 996 iterations.
My score = 0.5001758194232604
Model fit: elapsed = 40.418 with test/NDCG@10 = 0.5001758194232604
-------------------- Model number 20 --------------------
Params:
	 - learning_rate: 0.02607586865143843
	 - depth: 4
	 - l2_leaf_reg: 14.523012412832882
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.7733276776004808
	 - use_best_model: True
	 - min_data_in_leaf: 94
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_type: GP

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3003750	best: 0.3003750 (0)	total: 44.1ms	remaining: 44s
200:	test: 0.4473293	best: 0.4473293 (200)	total: 5.63s	remaining: 22.4s
400:	test: 0.4640520	best: 0.4640520 (400)	total: 11.2s	remaining: 16.8s
600:	test: 0.4754400	best: 0.4754400 (600)	total: 16.9s	remaining: 11.2s
800:	test: 0.4829184	best: 0.4829936 (790)	total: 22.4s	remaining: 5.57s
999:	test: 0.4869265	best: 0.4869265 (999)	total: 27.9s	remaining: 0us
bestTest = 0.4869265217
bestIteration = 999
My score = 0.4869265216696157
Model fit: elapsed = 31.822 with test/NDCG@10 = 0.4869265216696157
-------------------- Model number 21 --------------------
Params:
	 - learning_rate: 0.016737988780906453
	 - depth: 8
	 - l2_leaf_reg: 4.200879093764849
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.896898876839582
	 - use_best_model: True
	 - min_data_in_leaf: 9
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3373056	best: 0.3373056 (0)	total: 50.3ms	remaining: 50.3s
200:	test: 0.4367181	best: 0.4367181 (200)	total: 6.79s	remaining: 27s
400:	test: 0.4560709	best: 0.4560709 (400)	total: 13.5s	remaining: 20.2s
600:	test: 0.4667287	best: 0.4667287 (600)	total: 20.3s	remaining: 13.5s
800:	test: 0.4737140	best: 0.4738687 (785)	total: 27s	remaining: 6.71s
999:	test: 0.4809564	best: 0.4809564 (999)	total: 33.6s	remaining: 0us
bestTest = 0.4809564445
bestIteration = 999
My score = 0.48095644454911046
Model fit: elapsed = 36.112 with test/NDCG@10 = 0.48095644454911046
-------------------- Model number 22 --------------------
Params:
	 - learning_rate: 0.014680112183512766
	 - depth: 7
	 - l2_leaf_reg: 6.700888488721551
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8106175511982892
	 - use_best_model: True
	 - min_data_in_leaf: 29
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3045767	best: 0.3045767 (0)	total: 47.9ms	remaining: 47.9s
200:	test: 0.4280112	best: 0.4280112 (200)	total: 6.35s	remaining: 25.2s
400:	test: 0.4466759	best: 0.4466759 (400)	total: 12.7s	remaining: 19s
600:	test: 0.4581589	best: 0.4581589 (600)	total: 19s	remaining: 12.6s
800:	test: 0.4657254	best: 0.4657254 (800)	total: 25.2s	remaining: 6.26s
999:	test: 0.4713013	best: 0.4713013 (999)	total: 31.4s	remaining: 0us
bestTest = 0.4713012549
bestIteration = 999
My score = 0.4713012548500634
Model fit: elapsed = 33.879 with test/NDCG@10 = 0.4713012548500634
-------------------- Model number 23 --------------------
Params:
	 - learning_rate: 0.08227832975494558
	 - depth: 4
	 - l2_leaf_reg: 3.8159369979987767
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.9513834722496411
	 - use_best_model: True
	 - min_data_in_leaf: 70
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_type: GPU


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3004144	best: 0.3004144 (0)	total: 43.7ms	remaining: 43.6s
200:	test: 0.4785926	best: 0.4785926 (200)	total: 5.59s	remaining: 22.2s
400:	test: 0.4915215	best: 0.4915850 (395)	total: 11.2s	remaining: 16.7s
600:	test: 0.4959085	best: 0.4961803 (585)	total: 16.6s	remaining: 11s
800:	test: 0.4998378	best: 0.4999489 (765)	total: 22s	remaining: 5.47s
999:	test: 0.5014655	best: 0.5015344 (990)	total: 27.4s	remaining: 0us
bestTest = 0.5015344185
bestIteration = 990
Shrink model to first 991 iterations.
My score = 0.5015344185307874
Model fit: elapsed = 31.630 with test/NDCG@10 = 0.5015344185307874
-------------------- Model number 24 --------------------
Params:
	 - learning_rate: 0.010513742381502343
	 - depth: 3
	 - l2_leaf_reg: 29.317324485388475
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.8405953604943104
	 - use_best_model: True
	 - min_data_in_leaf: 11
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_typ

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2727262	best: 0.2727262 (0)	total: 42.9ms	remaining: 42.9s
200:	test: 0.4130109	best: 0.4130109 (200)	total: 5.21s	remaining: 20.7s
400:	test: 0.4283201	best: 0.4283201 (400)	total: 10.5s	remaining: 15.7s
600:	test: 0.4362267	best: 0.4362267 (600)	total: 15.7s	remaining: 10.4s
800:	test: 0.4447355	best: 0.4447934 (795)	total: 20.9s	remaining: 5.19s
999:	test: 0.4507770	best: 0.4507994 (995)	total: 26.1s	remaining: 0us
bestTest = 0.4507994306
bestIteration = 995
Shrink model to first 996 iterations.
My score = 0.4507994305564217
Model fit: elapsed = 29.638 with test/NDCG@10 = 0.4507994305564217
-------------------- Model number 25 --------------------
Params:
	 - learning_rate: 0.09769934773084338
	 - depth: 7
	 - l2_leaf_reg: 18.540520072606334
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 1000
	 - bagging_temperature: 0.9217790738194904
	 - use_best_model: True
	 - min_data_in_leaf: 63
	 - verbose: 200
	 - grow_policy: SymmetricTree
	 - t

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2922021	best: 0.2922021 (0)	total: 48ms	remaining: 47.9s
200:	test: 0.4737007	best: 0.4737007 (200)	total: 6.32s	remaining: 25.1s
400:	test: 0.4906803	best: 0.4908657 (390)	total: 12.6s	remaining: 18.9s
600:	test: 0.4963194	best: 0.4963194 (600)	total: 18.9s	remaining: 12.5s
800:	test: 0.4989511	best: 0.4989556 (795)	total: 25.1s	remaining: 6.23s
999:	test: 0.5017963	best: 0.5018639 (995)	total: 31.3s	remaining: 0us
bestTest = 0.5018638519
bestIteration = 995
Shrink model to first 996 iterations.
My score = 0.5018638518645737
Model fit: elapsed = 33.753 with test/NDCG@10 = 0.5018638518645737


Попробуем еще немного подкрутить параметры:

In [59]:
def delta_params(n_iter, params, pool_train, pool_valid):
    np.random.seed(0)
    random.seed(0)
    best_score = 0
    best_model = None
    best_params = {}
    y_valid, q_valid = pool_valid.get_label(), pool_valid.get_group_id_hash()

    for i in range(n_iter):
        print(f'{"-"*20} Model number {i+1} {"-"*20}')
        params['learning_rate'] = params['learning_rate'] + np.random.normal(0, 0.01)
        params['depth'] = random.randint(3, 8)
        params['l2_leaf_reg'] = params['l2_leaf_reg'] + np.random.normal(0, 3)
        params['iterations'] = 5000
        params['bagging_temperature'] = params['bagging_temperature'] + np.random.uniform(0, 0.05)
        params['min_data_in_leaf'] = random.randint(40, 70)
        params['random_seed'] = 0
        
        print('Params:')
        for k,v in params.items():
            print(f'\t - {k}: {v}')
        model = catboost.CatBoost(params)
        start = timer()
        model.fit(pool_train, eval_set=pool_valid)
        score = utils.eval_metric(y_valid, model.predict(pool_valid), 
                                  'NDCG:top=10;type=Exp', group_id=q_valid)[0]        
        if score > best_score:
            best_model = model
            best_score = score
            best_params = params
            
        elapsed = timer() - start
        print(f"Model fit: elapsed = {elapsed:.3f} with test/NDCG@10 = {score}")
        
    return best_model, best_params

In [60]:
best_model, best_params = delta_params(10, best_params, pool_train, pool_valid)

-------------------- Model number 1 --------------------
Params:
	 - learning_rate: 0.08614073307092152
	 - depth: 6
	 - l2_leaf_reg: 12.626200741741464
	 - loss_function: YetiRank
	 - eval_metric: NDCG:top=10;type=Exp
	 - iterations: 5000
	 - bagging_temperature: 0.8612477549433846
	 - use_best_model: True
	 - min_data_in_leaf: 64
	 - verbose: 200
	 - grow_policy: Depthwise
	 - task_type: GPU
	 - random_seed: 0


Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3375796	best: 0.3375796 (0)	total: 51.3ms	remaining: 4m 16s
200:	test: 0.4914603	best: 0.4914603 (200)	total: 6.39s	remaining: 2m 32s
400:	test: 0.5030810	best: 0.5030810 (400)	total: 12.7s	remaining: 2m 25s
600:	test: 0.5066836	best: 0.5067352 (595)	total: 18.9s	remaining: 2m 18s
800:	test: 0.5087713	best: 0.5088853 (795)	total: 25.2s	remaining: 2m 11s
1000:	test: 0.5094321	best: 0.5095146 (980)	total: 31.3s	remaining: 2m 4s
1200:	test: 0.5103396	best: 0.5104370 (1180)	total: 37.4s	remaining: 1m 58s
1400:	test: 0.5113061	best: 0.5113061 (1400)	total: 43.5s	remaining: 1m 51s
1600:	test: 0.5116712	best: 0.5116712 (1600)	total: 49.6s	remaining: 1m 45s
1800:	test: 0.5124676	best: 0.5124763 (1785)	total: 55.8s	remaining: 1m 39s
2000:	test: 0.5130439	best: 0.5130844 (1995)	total: 1m 1s	remaining: 1m 32s
2200:	test: 0.5137877	best: 0.5137890 (2175)	total: 1m 7s	remaining: 1m 26s
2400:	test: 0.5142659	best: 0.5144703 (2330)	total: 1m 14s	remaining: 1m 20s
2600:	test: 0.5145068	best

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3276757	best: 0.3276757 (0)	total: 50.4ms	remaining: 4m 11s
200:	test: 0.4842210	best: 0.4842210 (200)	total: 6.47s	remaining: 2m 34s
400:	test: 0.4982201	best: 0.4982201 (400)	total: 12.9s	remaining: 2m 27s
600:	test: 0.5034078	best: 0.5034078 (600)	total: 19.3s	remaining: 2m 21s
800:	test: 0.5056376	best: 0.5057383 (795)	total: 25.5s	remaining: 2m 13s
1000:	test: 0.5076827	best: 0.5076827 (1000)	total: 31.6s	remaining: 2m 6s
1200:	test: 0.5086811	best: 0.5086811 (1200)	total: 37.8s	remaining: 1m 59s
1400:	test: 0.5097896	best: 0.5097896 (1400)	total: 44s	remaining: 1m 53s
1600:	test: 0.5106804	best: 0.5107442 (1580)	total: 50.2s	remaining: 1m 46s
1800:	test: 0.5110365	best: 0.5112940 (1765)	total: 56.3s	remaining: 1m 40s
2000:	test: 0.5117474	best: 0.5118484 (1945)	total: 1m 2s	remaining: 1m 33s
2200:	test: 0.5124994	best: 0.5125776 (2180)	total: 1m 8s	remaining: 1m 27s
2400:	test: 0.5127427	best: 0.5128292 (2345)	total: 1m 14s	remaining: 1m 20s
2600:	test: 0.5129417	best:

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3161235	best: 0.3161235 (0)	total: 47.8ms	remaining: 3m 58s
200:	test: 0.4808077	best: 0.4808077 (200)	total: 6.04s	remaining: 2m 24s
400:	test: 0.4953857	best: 0.4953857 (400)	total: 12s	remaining: 2m 17s
600:	test: 0.5002070	best: 0.5002070 (600)	total: 17.8s	remaining: 2m 9s
800:	test: 0.5029828	best: 0.5030174 (795)	total: 23.5s	remaining: 2m 3s
1000:	test: 0.5048911	best: 0.5049671 (995)	total: 29.3s	remaining: 1m 56s
1200:	test: 0.5062596	best: 0.5063404 (1195)	total: 35s	remaining: 1m 50s
1400:	test: 0.5069554	best: 0.5072104 (1380)	total: 40.8s	remaining: 1m 44s
1600:	test: 0.5080728	best: 0.5081444 (1585)	total: 46.5s	remaining: 1m 38s
1800:	test: 0.5086471	best: 0.5090137 (1765)	total: 52.2s	remaining: 1m 32s
2000:	test: 0.5093757	best: 0.5094496 (1990)	total: 57.9s	remaining: 1m 26s
2200:	test: 0.5095733	best: 0.5098660 (2070)	total: 1m 3s	remaining: 1m 20s
2400:	test: 0.5095708	best: 0.5098660 (2070)	total: 1m 9s	remaining: 1m 14s
2600:	test: 0.5101804	best: 0.51

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3386992	best: 0.3386992 (0)	total: 56.5ms	remaining: 4m 42s
200:	test: 0.4924852	best: 0.4924852 (200)	total: 7.1s	remaining: 2m 49s
400:	test: 0.5033815	best: 0.5033815 (400)	total: 14.2s	remaining: 2m 42s
600:	test: 0.5078970	best: 0.5080274 (595)	total: 21s	remaining: 2m 33s
800:	test: 0.5096257	best: 0.5096803 (780)	total: 27.6s	remaining: 2m 24s
1000:	test: 0.5110261	best: 0.5110261 (1000)	total: 34.2s	remaining: 2m 16s
1200:	test: 0.5118878	best: 0.5119230 (1195)	total: 41s	remaining: 2m 9s
1400:	test: 0.5132095	best: 0.5132649 (1385)	total: 47.6s	remaining: 2m 2s
1600:	test: 0.5141737	best: 0.5141737 (1600)	total: 54.2s	remaining: 1m 55s
1800:	test: 0.5149359	best: 0.5152500 (1765)	total: 1m	remaining: 1m 47s
2000:	test: 0.5150107	best: 0.5152500 (1765)	total: 1m 7s	remaining: 1m 40s
2200:	test: 0.5157213	best: 0.5157213 (2200)	total: 1m 14s	remaining: 1m 34s
2400:	test: 0.5155679	best: 0.5157213 (2200)	total: 1m 20s	remaining: 1m 27s
2600:	test: 0.5156843	best: 0.515

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3352075	best: 0.3352075 (0)	total: 51.4ms	remaining: 4m 17s
200:	test: 0.4922165	best: 0.4922165 (200)	total: 6.42s	remaining: 2m 33s
400:	test: 0.5036548	best: 0.5036548 (400)	total: 12.7s	remaining: 2m 25s
600:	test: 0.5068424	best: 0.5071374 (595)	total: 19s	remaining: 2m 19s
800:	test: 0.5087507	best: 0.5087507 (800)	total: 25.2s	remaining: 2m 12s
1000:	test: 0.5096901	best: 0.5097481 (995)	total: 31.3s	remaining: 2m 5s
1200:	test: 0.5101781	best: 0.5103618 (1150)	total: 37.4s	remaining: 1m 58s
1400:	test: 0.5110764	best: 0.5110764 (1400)	total: 43.5s	remaining: 1m 51s
1600:	test: 0.5117631	best: 0.5118591 (1560)	total: 49.7s	remaining: 1m 45s
1800:	test: 0.5118850	best: 0.5121181 (1670)	total: 55.7s	remaining: 1m 39s
2000:	test: 0.5128837	best: 0.5129611 (1995)	total: 1m 1s	remaining: 1m 32s
2200:	test: 0.5136773	best: 0.5136923 (2175)	total: 1m 7s	remaining: 1m 26s
2400:	test: 0.5136190	best: 0.5139261 (2345)	total: 1m 13s	remaining: 1m 20s
2600:	test: 0.5137916	best: 

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3191325	best: 0.3191325 (0)	total: 47.2ms	remaining: 3m 56s
200:	test: 0.4876881	best: 0.4876881 (200)	total: 5.92s	remaining: 2m 21s
400:	test: 0.4984276	best: 0.4984276 (400)	total: 11.9s	remaining: 2m 16s
600:	test: 0.5026452	best: 0.5026452 (600)	total: 17.7s	remaining: 2m 9s
800:	test: 0.5047631	best: 0.5050223 (795)	total: 23.5s	remaining: 2m 3s
1000:	test: 0.5066601	best: 0.5066601 (1000)	total: 29.2s	remaining: 1m 56s
1200:	test: 0.5081168	best: 0.5081168 (1200)	total: 34.9s	remaining: 1m 50s
1400:	test: 0.5085713	best: 0.5087864 (1395)	total: 40.8s	remaining: 1m 44s
1600:	test: 0.5089363	best: 0.5090192 (1430)	total: 46.5s	remaining: 1m 38s
1800:	test: 0.5095857	best: 0.5095897 (1795)	total: 52.2s	remaining: 1m 32s
2000:	test: 0.5098175	best: 0.5099391 (1875)	total: 57.9s	remaining: 1m 26s
2200:	test: 0.5106845	best: 0.5108488 (2165)	total: 1m 3s	remaining: 1m 20s
2400:	test: 0.5111346	best: 0.5111346 (2400)	total: 1m 9s	remaining: 1m 14s
2600:	test: 0.5112810	best:

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3340910	best: 0.3340910 (0)	total: 50.3ms	remaining: 4m 11s
200:	test: 0.4953969	best: 0.4953969 (200)	total: 6.41s	remaining: 2m 33s
400:	test: 0.5045058	best: 0.5046454 (390)	total: 12.7s	remaining: 2m 25s
600:	test: 0.5076771	best: 0.5076771 (600)	total: 18.8s	remaining: 2m 17s
800:	test: 0.5089869	best: 0.5091929 (770)	total: 24.9s	remaining: 2m 10s
1000:	test: 0.5102177	best: 0.5102177 (1000)	total: 31.1s	remaining: 2m 4s
1200:	test: 0.5105577	best: 0.5111726 (1125)	total: 37.2s	remaining: 1m 57s
1400:	test: 0.5113679	best: 0.5113794 (1350)	total: 43.4s	remaining: 1m 51s
1600:	test: 0.5124627	best: 0.5124627 (1600)	total: 49.5s	remaining: 1m 45s
1800:	test: 0.5120891	best: 0.5126200 (1615)	total: 55.6s	remaining: 1m 38s
2000:	test: 0.5128914	best: 0.5131038 (1980)	total: 1m 1s	remaining: 1m 32s
2200:	test: 0.5131447	best: 0.5132409 (2065)	total: 1m 7s	remaining: 1m 26s
2400:	test: 0.5132478	best: 0.5136653 (2290)	total: 1m 14s	remaining: 1m 20s
2600:	test: 0.5134354	bes

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3418100	best: 0.3418100 (0)	total: 54.7ms	remaining: 4m 33s
200:	test: 0.4961840	best: 0.4961840 (200)	total: 7.19s	remaining: 2m 51s
400:	test: 0.5039576	best: 0.5039576 (400)	total: 14.1s	remaining: 2m 41s
600:	test: 0.5074921	best: 0.5075773 (595)	total: 20.9s	remaining: 2m 33s
800:	test: 0.5097502	best: 0.5098805 (790)	total: 27.6s	remaining: 2m 24s
1000:	test: 0.5105890	best: 0.5111266 (920)	total: 34.4s	remaining: 2m 17s
1200:	test: 0.5117033	best: 0.5117033 (1200)	total: 41.1s	remaining: 2m 10s
1400:	test: 0.5129426	best: 0.5131209 (1390)	total: 47.7s	remaining: 2m 2s
1600:	test: 0.5137465	best: 0.5137483 (1595)	total: 54.3s	remaining: 1m 55s
1800:	test: 0.5139969	best: 0.5141513 (1745)	total: 1m	remaining: 1m 48s
2000:	test: 0.5143871	best: 0.5145600 (1885)	total: 1m 7s	remaining: 1m 41s
2200:	test: 0.5147630	best: 0.5149336 (2105)	total: 1m 14s	remaining: 1m 34s
2400:	test: 0.5151799	best: 0.5153257 (2355)	total: 1m 20s	remaining: 1m 27s
2600:	test: 0.5155247	best: 

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2856607	best: 0.2856607 (0)	total: 44.3ms	remaining: 3m 41s
200:	test: 0.4738129	best: 0.4738129 (200)	total: 5.56s	remaining: 2m 12s
400:	test: 0.4889841	best: 0.4889841 (400)	total: 11.1s	remaining: 2m 7s
600:	test: 0.4943728	best: 0.4943728 (600)	total: 16.6s	remaining: 2m 1s
800:	test: 0.4978224	best: 0.4978254 (795)	total: 22s	remaining: 1m 55s
1000:	test: 0.4991094	best: 0.4991094 (1000)	total: 27.4s	remaining: 1m 49s
1200:	test: 0.5011807	best: 0.5011807 (1200)	total: 32.8s	remaining: 1m 43s
1400:	test: 0.5018125	best: 0.5018125 (1400)	total: 38.2s	remaining: 1m 38s
1600:	test: 0.5026804	best: 0.5026804 (1600)	total: 43.7s	remaining: 1m 32s
1800:	test: 0.5033632	best: 0.5033632 (1800)	total: 49.2s	remaining: 1m 27s
2000:	test: 0.5037400	best: 0.5037400 (2000)	total: 54.7s	remaining: 1m 21s
2200:	test: 0.5043087	best: 0.5043087 (2200)	total: 60s	remaining: 1m 16s
2400:	test: 0.5043341	best: 0.5045145 (2360)	total: 1m 5s	remaining: 1m 10s
2600:	test: 0.5052252	best: 0.5

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2861461	best: 0.2861461 (0)	total: 43.5ms	remaining: 3m 37s
200:	test: 0.4717602	best: 0.4717602 (200)	total: 5.58s	remaining: 2m 13s
400:	test: 0.4871494	best: 0.4872725 (395)	total: 11.1s	remaining: 2m 6s
600:	test: 0.4934174	best: 0.4935417 (595)	total: 16.5s	remaining: 2m
800:	test: 0.4967148	best: 0.4969612 (760)	total: 21.9s	remaining: 1m 54s
1000:	test: 0.4993367	best: 0.4993367 (1000)	total: 27.3s	remaining: 1m 49s
1200:	test: 0.5004984	best: 0.5007992 (1175)	total: 32.7s	remaining: 1m 43s
1400:	test: 0.5018247	best: 0.5018247 (1400)	total: 38.1s	remaining: 1m 37s
1600:	test: 0.5020964	best: 0.5022449 (1565)	total: 43.5s	remaining: 1m 32s
1800:	test: 0.5023260	best: 0.5024984 (1685)	total: 48.8s	remaining: 1m 26s
2000:	test: 0.5030460	best: 0.5033603 (1980)	total: 54.4s	remaining: 1m 21s
2200:	test: 0.5036365	best: 0.5037122 (2190)	total: 59.7s	remaining: 1m 15s
2400:	test: 0.5047472	best: 0.5047472 (2400)	total: 1m 5s	remaining: 1m 10s
2600:	test: 0.5051235	best: 0.

In [64]:
best_model.get_params()

{'learning_rate': 0.10340626393986621,
 'depth': 7,
 'l2_leaf_reg': 17.272888261719643,
 'loss_function': 'YetiRank',
 'eval_metric': 'NDCG:top=10;type=Exp',
 'iterations': 5000,
 'bagging_temperature': 1.068369922077575,
 'use_best_model': True,
 'min_data_in_leaf': 68,
 'verbose': 200,
 'grow_policy': 'Depthwise',
 'task_type': 'GPU',
 'random_seed': 0}

### Сохраняем модель

In [65]:
with open('best_params.pkl', 'wb') as f:
    pickle.dump(best_model.get_params(), f)

### Загружаем модель

In [66]:
best_params = {}
with open('/kaggle/input/vk-catboost-params/other/catboost-vk-v3/1/best_params.pkl', 'rb') as f:
    best_params = pickle.load(f)

In [67]:
best_params

{'learning_rate': 0.10340626393986621,
 'depth': 7,
 'l2_leaf_reg': 17.272888261719643,
 'loss_function': 'YetiRank',
 'eval_metric': 'NDCG:top=10;type=Exp',
 'iterations': 5000,
 'bagging_temperature': 1.068369922077575,
 'use_best_model': True,
 'min_data_in_leaf': 68,
 'verbose': 200,
 'grow_policy': 'Depthwise',
 'task_type': 'GPU',
 'random_seed': 0}

### Проверяем качество + подберем оптимальное число итераций.

In [69]:
best_params['iterations'] = 10_000

model = catboost.CatBoost(best_params)
start = timer()
model.fit(pool_train, eval_set=pool_valid)

elapsed = timer() - start
print(f"Model fit: elapsed = {elapsed:.3f} with test/NDCG@10 = {score}")

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3418100	best: 0.3418100 (0)	total: 54.4ms	remaining: 9m 4s
200:	test: 0.4961840	best: 0.4961840 (200)	total: 7.08s	remaining: 5m 45s
400:	test: 0.5044202	best: 0.5044202 (400)	total: 14s	remaining: 5m 33s
600:	test: 0.5080357	best: 0.5080705 (595)	total: 20.7s	remaining: 5m 24s
800:	test: 0.5100536	best: 0.5100536 (800)	total: 27.4s	remaining: 5m 15s
1000:	test: 0.5115030	best: 0.5116374 (980)	total: 34.1s	remaining: 5m 6s
1200:	test: 0.5119654	best: 0.5122434 (1065)	total: 40.8s	remaining: 4m 58s
1400:	test: 0.5129246	best: 0.5129774 (1395)	total: 47.4s	remaining: 4m 51s
1600:	test: 0.5131108	best: 0.5134650 (1520)	total: 54s	remaining: 4m 43s
1800:	test: 0.5136245	best: 0.5138132 (1655)	total: 1m	remaining: 4m 35s
2000:	test: 0.5143435	best: 0.5143435 (2000)	total: 1m 7s	remaining: 4m 28s
2200:	test: 0.5152589	best: 0.5153157 (2185)	total: 1m 13s	remaining: 4m 21s
2400:	test: 0.5154514	best: 0.5158513 (2260)	total: 1m 20s	remaining: 4m 14s
2600:	test: 0.5155680	best: 0.515

Оптимальным числом итераций является 7775. NDSG@10 на валидаци 0.519

## Проведем feature selection.

### Baseline 1000 итераций 100% признаков

In [18]:
best_params['iterations'] = 1000
model = catboost.CatBoost(best_params)
start = timer()
model.fit(pool_train, eval_set=pool_valid)
y_pred = model.predict(X_test)
score = utils.eval_metric(y_test, y_pred, 'NDCG:top=10;type=Exp', group_id=q_test)[0]

elapsed = timer() - start
print(f"Model fit: elapsed = {elapsed:.3f} with test/NDCG@10 = {score}")

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3418100	best: 0.3418100 (0)	total: 21.6s	remaining: 5h 59m 31s
200:	test: 0.4961840	best: 0.4961840 (200)	total: 28.7s	remaining: 1m 53s
400:	test: 0.5039576	best: 0.5039576 (400)	total: 35.6s	remaining: 53.2s
600:	test: 0.5075725	best: 0.5075725 (600)	total: 42.5s	remaining: 28.2s
800:	test: 0.5098837	best: 0.5098837 (800)	total: 49.3s	remaining: 12.3s
999:	test: 0.5107292	best: 0.5110378 (975)	total: 56s	remaining: 0us
bestTest = 0.5110378479
bestIteration = 975
Shrink model to first 976 iterations.
Model fit: elapsed = 62.137 with test/NDCG@10 = 0.5038919620562503


### 75% признаков

In [19]:
result75 = model.select_features(
                pool_train,
                eval_set=pool_valid,
                features_for_select=np.arange(X_train.shape[1]),
                num_features_to_select = 3 * X_train.shape[1] // 4,
                algorithm='RecursiveByShapValues',
                steps=1,
                shap_calc_type='Approximate',
                train_final_model=False)['selected_features']

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


Step #1 out of 1
0:	test: 0.3418100	best: 0.3418100 (0)	total: 55.5ms	remaining: 55.5s
200:	test: 0.4961840	best: 0.4961840 (200)	total: 7.2s	remaining: 28.6s
400:	test: 0.5039576	best: 0.5039576 (400)	total: 14.2s	remaining: 21.2s
600:	test: 0.5075707	best: 0.5075707 (600)	total: 21.1s	remaining: 14s
800:	test: 0.5097039	best: 0.5097896 (790)	total: 27.9s	remaining: 6.93s
999:	test: 0.5112119	best: 0.5112305 (975)	total: 34.6s	remaining: 0us
bestTest = 0.5112304998
bestIteration = 975
Shrink model to first 976 iterations.
Feature #121 eliminated
Feature #90 eliminated
Feature #94 eliminated
Feature #106 eliminated
Feature #76 eliminated
Feature #44 eliminated
Feature #43 eliminated
Feature #0 eliminated
Feature #51 eliminated
Feature #27 eliminated
Feature #31 eliminated
Feature #61 eliminated
Feature #33 eliminated
Feature #28 eliminated
Feature #96 eliminated
Feature #97 eliminated
Feature #98 eliminated
Feature #99 eliminated
Feature #36 eliminated
Feature #104 eliminated
Feature #

Проверим качество:

In [20]:
pool_train_selected = catboost.Pool(data=X_train[:, result75], label=y_train, group_id=q_train)
pool_valid_selected = catboost.Pool(data=X_valid[:, result75], label=y_valid, group_id=q_valid)
X_test_selected = X_test[:, result75]
X_test_selected.shape

(241521, 102)

In [21]:
best_params['iterations'] = 10_000
model1 = catboost.CatBoost(best_params)
start = timer()
model1.fit(pool_train_selected, eval_set=pool_valid_selected)
y_pred = model1.predict(X_test_selected)
score = utils.eval_metric(y_test, y_pred, 'NDCG:top=10;type=Exp', group_id=q_test)[0]

elapsed = timer() - start
print(f"Model fit: elapsed = {elapsed:.3f} with test/NDCG@10 = {score}")

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3401737	best: 0.3401737 (0)	total: 49.9ms	remaining: 8m 18s
200:	test: 0.4964417	best: 0.4964417 (200)	total: 6.86s	remaining: 5m 34s
400:	test: 0.5042694	best: 0.5042694 (400)	total: 13.5s	remaining: 5m 23s
600:	test: 0.5088452	best: 0.5088452 (600)	total: 20.1s	remaining: 5m 14s
800:	test: 0.5100050	best: 0.5100050 (800)	total: 26.6s	remaining: 5m 5s
1000:	test: 0.5108595	best: 0.5110758 (990)	total: 33.1s	remaining: 4m 57s
1200:	test: 0.5125296	best: 0.5126890 (1175)	total: 39.6s	remaining: 4m 50s
1400:	test: 0.5132054	best: 0.5134765 (1305)	total: 46s	remaining: 4m 42s
1600:	test: 0.5136946	best: 0.5140209 (1530)	total: 52.6s	remaining: 4m 36s
1800:	test: 0.5143327	best: 0.5143907 (1785)	total: 59.1s	remaining: 4m 29s
2000:	test: 0.5144962	best: 0.5144983 (1985)	total: 1m 5s	remaining: 4m 22s
2200:	test: 0.5152687	best: 0.5152687 (2200)	total: 1m 12s	remaining: 4m 15s
2400:	test: 0.5152965	best: 0.5154664 (2230)	total: 1m 18s	remaining: 4m 8s
2600:	test: 0.5156696	best: 

### 50% признаков

In [22]:
result50 = model.select_features(
                pool_train,
                eval_set=pool_valid,
                features_for_select=np.arange(X_train.shape[1]),
                num_features_to_select = X_train.shape[1] // 2,
                algorithm='RecursiveByShapValues',
                steps=1,
                shap_calc_type='Approximate',
                train_final_model=False)['selected_features']

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


Step #1 out of 1
0:	test: 0.3418100	best: 0.3418100 (0)	total: 55.2ms	remaining: 55.1s
200:	test: 0.4961840	best: 0.4961840 (200)	total: 7.13s	remaining: 28.3s
400:	test: 0.5039576	best: 0.5039576 (400)	total: 14.1s	remaining: 21.1s
600:	test: 0.5075725	best: 0.5075725 (600)	total: 20.9s	remaining: 13.9s
800:	test: 0.5095598	best: 0.5096511 (785)	total: 27.6s	remaining: 6.87s
999:	test: 0.5109636	best: 0.5111249 (945)	total: 34.3s	remaining: 0us
bestTest = 0.5111248997
bestIteration = 945
Shrink model to first 946 iterations.
Feature #100 eliminated
Feature #83 eliminated
Feature #46 eliminated
Feature #22 eliminated
Feature #76 eliminated
Feature #95 eliminated
Feature #81 eliminated
Feature #61 eliminated
Feature #37 eliminated
Feature #79 eliminated
Feature #28 eliminated
Feature #26 eliminated
Feature #31 eliminated
Feature #96 eliminated
Feature #97 eliminated
Feature #98 eliminated
Feature #99 eliminated
Feature #56 eliminated
Feature #0 eliminated
Feature #33 eliminated
Feature 

In [23]:
pool_train_selected = catboost.Pool(data=X_train[:, result50], label=y_train, group_id=q_train)
pool_valid_selected = catboost.Pool(data=X_valid[:, result50], label=y_valid, group_id=q_valid)
X_test_selected = X_test[:, result50]
X_test_selected.shape

(241521, 68)

In [24]:
best_params['iterations'] = 10_000
model1 = catboost.CatBoost(best_params)
start = timer()
model1.fit(pool_train_selected, eval_set=pool_valid_selected)
y_pred = model1.predict(X_test_selected)
score = utils.eval_metric(y_test, y_pred, 'NDCG:top=10;type=Exp', group_id=q_test)[0]

elapsed = timer() - start
print(f"Model fit: elapsed = {elapsed:.3f} with test/NDCG@10 = {score}")

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3447758	best: 0.3447758 (0)	total: 50.5ms	remaining: 8m 25s
200:	test: 0.4945795	best: 0.4945795 (200)	total: 6.58s	remaining: 5m 20s
400:	test: 0.5045132	best: 0.5045763 (380)	total: 13s	remaining: 5m 10s
600:	test: 0.5078777	best: 0.5078777 (600)	total: 19.4s	remaining: 5m 2s
800:	test: 0.5097308	best: 0.5099568 (780)	total: 25.7s	remaining: 4m 55s
1000:	test: 0.5108019	best: 0.5108547 (940)	total: 32s	remaining: 4m 47s
1200:	test: 0.5111183	best: 0.5111188 (1045)	total: 38.3s	remaining: 4m 40s
1400:	test: 0.5118976	best: 0.5118976 (1400)	total: 44.6s	remaining: 4m 33s
1600:	test: 0.5127989	best: 0.5127989 (1600)	total: 50.9s	remaining: 4m 26s
1800:	test: 0.5129098	best: 0.5133581 (1755)	total: 57.3s	remaining: 4m 20s
2000:	test: 0.5132511	best: 0.5133581 (1755)	total: 1m 3s	remaining: 4m 13s
2200:	test: 0.5137977	best: 0.5139482 (2190)	total: 1m 9s	remaining: 4m 6s
2400:	test: 0.5141027	best: 0.5142811 (2350)	total: 1m 15s	remaining: 4m
2600:	test: 0.5140284	best: 0.51436

Для дальнейших экспериментов, чтобы хватало оперативной памяти, сохраним половину отобранных признаков.

In [25]:
with open('best_indeces.pkl', 'wb') as f:
    pickle.dump(result50, f)

In [27]:
result50[:20]

[1, 2, 5, 7, 10, 11, 12, 13, 14, 17, 19, 20, 24, 29, 35, 40, 45, 47, 48, 49]

In [28]:
with open('/kaggle/input/vk-catboost-params/other/half_features/1/best_indeces.pkl', 'rb') as f:
    result50 = pickle.load(f)
result50[:20]

[1, 2, 5, 7, 10, 11, 12, 13, 14, 17, 19, 20, 24, 29, 35, 40, 45, 47, 48, 49]

In [29]:
X_train_selected = X_train[:, result50]
X_valid_selected = X_valid[:, result50]
X_test_selected = X_test[:, result50]
X_train_selected.shape, X_valid_selected.shape, X_test_selected.shape

((723412, 68), (235259, 68), (241521, 68))

Видно, что качество ухудшилось, значит оставим 75% признаков.

### Попробуем добавить примитивную кластеризацию.

За признаки возьмем растояния до центров классов.

In [16]:
from sklearn.cluster import KMeans

In [17]:
def euclidean_distance(X: np.array, Y: np.array) -> np.array:
    n, m = X.shape[0], Y.shape[0]

    Z = np.dot(X, Y.T)
    X = np.einsum('ij,ij->i', X, X).reshape(n, 1)
    Y = np.einsum('ij,ij->i', Y, Y)

    X = np.repeat(X, repeats=m, axis=1)
    Y = np.tile(Y, (n, 1))

    return (X - 2 * Z + Y) 

In [18]:
clusterts_dict = {}
for n_clusters in [64, 128, 256, 512, 1024]:
    print(f'{"-"*20} KMeans k = {n_clusters} {"-"*20}')
    kmeans = KMeans(n_clusters=n_clusters, n_init=3, verbose=1, random_state=0)
    kmeans.fit(X_train)
    
    clusters = kmeans.cluster_centers_
    clusterts_dict[n_clusters] = clusters
    
    X_train_distances = euclidean_distance(X_train, clusters)
    X_valid_distances = euclidean_distance(X_valid, clusters)
    pool_train_distances = catboost.Pool(data=X_train_distances, label=y_train, group_id=q_train)
    pool_valid_distances = catboost.Pool(data=X_valid_distances, label=y_valid, group_id=q_valid)
    
    X_test_distances = euclidean_distance(X_test, clusters)

    best_params['iterations'] = 1000
    model = catboost.CatBoost(best_params)
    start = timer()
    model.fit(pool_train_distances, eval_set=pool_valid_distances)
    y_pred = model.predict(X_test_distances)
    score = utils.eval_metric(y_test, y_pred, 'NDCG:top=10;type=Exp', group_id=q_test)[0]

-------------------- KMeans k = 64 --------------------
Initialization complete
Iteration 0, inertia 1025100379848704.0.
Iteration 1, inertia 889565170630656.0.
Iteration 2, inertia 871214318878720.0.
Iteration 3, inertia 861057392312320.0.
Iteration 4, inertia 860232221720576.0.
Iteration 5, inertia 859795208798208.0.
Iteration 6, inertia 859429733924864.0.
Iteration 7, inertia 859300012490752.0.
Converged at iteration 7: center shift 23859972.0 within tolerance 50386282.0864.
Initialization complete
Iteration 0, inertia 1032960371326976.0.
Iteration 1, inertia 851611349942272.0.
Iteration 2, inertia 838175014518784.0.
Iteration 3, inertia 832582363119616.0.
Iteration 4, inertia 830094905966592.0.
Iteration 5, inertia 829118606213120.0.
Iteration 6, inertia 828359739179008.0.
Iteration 7, inertia 827901586964480.0.
Iteration 8, inertia 827615434768384.0.
Iteration 9, inertia 827485042245632.0.
Iteration 10, inertia 827299754672128.0.
Iteration 11, inertia 827198218960896.0.
Iteration 

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2610270	best: 0.2610270 (0)	total: 16.8s	remaining: 4h 39m 52s
200:	test: 0.3409517	best: 0.3409517 (200)	total: 23.2s	remaining: 1m 32s
400:	test: 0.3487591	best: 0.3487591 (400)	total: 29.4s	remaining: 44s
600:	test: 0.3527338	best: 0.3527338 (600)	total: 35.7s	remaining: 23.7s
800:	test: 0.3542855	best: 0.3545040 (775)	total: 41.8s	remaining: 10.4s
999:	test: 0.3552373	best: 0.3555762 (975)	total: 48s	remaining: 0us
bestTest = 0.3555761692
bestIteration = 975
Shrink model to first 976 iterations.
-------------------- KMeans k = 128 --------------------
Initialization complete
Iteration 0, inertia 146688836632576.0.
Iteration 1, inertia 122472192016384.0.
Iteration 2, inertia 119731407290368.0.
Iteration 3, inertia 118662698631168.0.
Iteration 4, inertia 118193616060416.0.
Iteration 5, inertia 117906364956672.0.
Converged at iteration 5: center shift 40671312.0 within tolerance 50386282.0864.
Initialization complete
Iteration 0, inertia 150017352400896.0.
Iteration 1, iner

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2484289	best: 0.2484289 (0)	total: 53.3ms	remaining: 53.2s
200:	test: 0.3522061	best: 0.3522061 (200)	total: 6.86s	remaining: 27.3s
400:	test: 0.3607850	best: 0.3608647 (395)	total: 13.5s	remaining: 20.1s
600:	test: 0.3645366	best: 0.3645366 (600)	total: 20s	remaining: 13.3s
800:	test: 0.3660644	best: 0.3661860 (780)	total: 26.6s	remaining: 6.6s
999:	test: 0.3670219	best: 0.3670389 (995)	total: 33.1s	remaining: 0us
bestTest = 0.3670389439
bestIteration = 995
Shrink model to first 996 iterations.
-------------------- KMeans k = 256 --------------------
Initialization complete
Iteration 0, inertia 42652670099456.0.
Iteration 1, inertia 35315779633152.0.
Iteration 2, inertia 34611157532672.0.
Iteration 3, inertia 34388700037120.0.
Iteration 4, inertia 34258567561216.0.
Iteration 5, inertia 34169860128768.0.
Iteration 6, inertia 34104483512320.0.
Iteration 7, inertia 34054040715264.0.
Converged at iteration 7: center shift 35053884.0 within tolerance 50386282.0864.
Initializatio

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2491225	best: 0.2491225 (0)	total: 60.3ms	remaining: 1m
200:	test: 0.3532263	best: 0.3532263 (200)	total: 7.74s	remaining: 30.8s
400:	test: 0.3604871	best: 0.3606508 (395)	total: 15.3s	remaining: 22.9s
600:	test: 0.3650180	best: 0.3652990 (575)	total: 22.7s	remaining: 15.1s
800:	test: 0.3665205	best: 0.3666187 (795)	total: 30.1s	remaining: 7.48s
999:	test: 0.3674001	best: 0.3675733 (925)	total: 37.4s	remaining: 0us
bestTest = 0.367573278
bestIteration = 925
Shrink model to first 926 iterations.
-------------------- KMeans k = 512 --------------------
Initialization complete
Iteration 0, inertia 18164362510336.0.
Iteration 1, inertia 15175381942272.0.
Iteration 2, inertia 14800046260224.0.
Iteration 3, inertia 14632655781888.0.
Iteration 4, inertia 14522331955200.0.
Iteration 5, inertia 14455513546752.0.
Iteration 6, inertia 14408683094016.0.
Iteration 7, inertia 14373696307200.0.
Converged at iteration 7: center shift 34070800.0 within tolerance 50386282.0864.
Initialization

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2578344	best: 0.2578344 (0)	total: 72.9ms	remaining: 1m 12s
200:	test: 0.3526963	best: 0.3526963 (200)	total: 9.45s	remaining: 37.6s
400:	test: 0.3610444	best: 0.3610444 (400)	total: 18.3s	remaining: 27.3s
600:	test: 0.3666168	best: 0.3666168 (600)	total: 27s	remaining: 18s
800:	test: 0.3693729	best: 0.3693729 (800)	total: 35.8s	remaining: 8.9s
999:	test: 0.3703651	best: 0.3704395 (960)	total: 44.6s	remaining: 0us
bestTest = 0.3704394685
bestIteration = 960
Shrink model to first 961 iterations.
-------------------- KMeans k = 1024 --------------------
Initialization complete
Iteration 0, inertia 9092598857728.0.
Iteration 1, inertia 7665983946752.0.
Iteration 2, inertia 7473280843776.0.
Iteration 3, inertia 7388425879552.0.
Iteration 4, inertia 7329665777664.0.
Iteration 5, inertia 7291101773824.0.
Iteration 6, inertia 7265859403776.0.
Iteration 7, inertia 7245778649088.0.
Iteration 8, inertia 7228815835136.0.
Iteration 9, inertia 7215057993728.0.
Iteration 10, inertia 72044

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2545986	best: 0.2545986 (0)	total: 91ms	remaining: 1m 30s
200:	test: 0.3553808	best: 0.3553808 (200)	total: 12.3s	remaining: 48.9s
400:	test: 0.3611496	best: 0.3611496 (400)	total: 23.9s	remaining: 35.6s
600:	test: 0.3665970	best: 0.3665970 (600)	total: 35.3s	remaining: 23.4s
800:	test: 0.3686452	best: 0.3687920 (780)	total: 46.8s	remaining: 11.6s
999:	test: 0.3704560	best: 0.3704560 (999)	total: 58.2s	remaining: 0us
bestTest = 0.3704560336
bestIteration = 999


In [37]:
clusterts_dict[128].shape

(128, 136)

In [38]:
X_train_distances = euclidean_distance(X_train, clusterts_dict[128])
X_valid_distances = euclidean_distance(X_valid, clusterts_dict[128])
X_train_distances.shape, X_valid_distances.shape

((723412, 128), (235259, 128))

In [39]:
pool_train_distances = catboost.Pool(
            data=np.hstack([X_train, X_train_distances]), 
            label=y_train, group_id=q_train
            )
pool_valid_distances = catboost.Pool(
            data=np.hstack([X_valid, X_valid_distances]),
            label=y_valid, group_id=q_valid
            )

X_test_distances = np.hstack(
    [X_test, euclidean_distance(X_test, clusterts_dict[128])]
   )

Проверим, произойдет ли прирост качества, и заодно найдем оптимальное число итераций.

In [40]:
best_params['iterations'] = 15_000
best_params['verbose'] = 1000
model = catboost.CatBoost(best_params)
start = timer()
model.fit(pool_train_distances, eval_set=pool_valid_distances)
y_pred = model.predict(X_test_distances)
score = utils.eval_metric(y_test, y_pred, 'NDCG:top=10;type=Exp', group_id=q_test)[0]

elapsed = timer() - start
print(f"Model fit: elapsed = {elapsed:.3f} with test/NDCG@10 = {score}")

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3448433	best: 0.3448433 (0)	total: 63.2ms	remaining: 15m 47s
1000:	test: 0.5103019	best: 0.5107394 (880)	total: 39.6s	remaining: 9m 14s
2000:	test: 0.5137487	best: 0.5140701 (1975)	total: 1m 17s	remaining: 8m 24s
3000:	test: 0.5151977	best: 0.5153651 (2950)	total: 1m 55s	remaining: 7m 42s
4000:	test: 0.5170166	best: 0.5170166 (4000)	total: 2m 33s	remaining: 7m 3s
5000:	test: 0.5177565	best: 0.5179227 (4980)	total: 3m 12s	remaining: 6m 24s
6000:	test: 0.5180898	best: 0.5184100 (5630)	total: 3m 50s	remaining: 5m 45s
7000:	test: 0.5186503	best: 0.5190027 (6960)	total: 4m 28s	remaining: 5m 7s
8000:	test: 0.5184952	best: 0.5190389 (7360)	total: 5m 7s	remaining: 4m 28s
9000:	test: 0.5181638	best: 0.5190389 (7360)	total: 5m 45s	remaining: 3m 50s
10000:	test: 0.5178333	best: 0.5190389 (7360)	total: 6m 24s	remaining: 3m 12s
11000:	test: 0.5180812	best: 0.5190389 (7360)	total: 7m 2s	remaining: 2m 33s
12000:	test: 0.5179128	best: 0.5190389 (7360)	total: 7m 41s	remaining: 1m 55s
13000:	

Итого, перебрав вручную k=64, 256, 128, 512 - оптимальное качество на валидации достигается при k=128. Но при этом качество на тесте ухудшается. Таким образом метод не дает никаких полезных результатов.

## Добавим вторые степени признаков.

In [30]:
X_train_with_squares = (X_train_selected[:,:,None]*X_train_selected[:,None,:]).reshape(X_train_selected.shape[0],-1)
X_valid_with_squares = (X_valid_selected[:,:,None]*X_valid_selected[:,None,:]).reshape(X_valid_selected.shape[0],-1)
X_test_with_squares = (X_test_selected[:,:,None]*X_test_selected[:,None,:]).reshape(X_test_selected.shape[0],-1)
X_train_with_squares.shape, X_valid_with_squares.shape, X_test_with_squares.shape

((723412, 4624), (235259, 4624), (241521, 4624))

In [32]:
pool_train_squares = catboost.Pool(data=X_train_with_squares, label=y_train, group_id=q_train)
pool_valid_squares = catboost.Pool(data=X_valid_with_squares, label=y_valid, group_id=q_valid)

In [33]:
best_params['iterations'] = 10_000
model1 = catboost.CatBoost(best_params)
start = timer()
model1.fit(pool_train_squares, eval_set=pool_valid_squares)
y_pred = model1.predict(X_test_with_squares)
score = utils.eval_metric(y_test, y_pred, 'NDCG:top=10;type=Exp', group_id=q_test)[0]

elapsed = timer() - start
print(f"Model fit: elapsed = {elapsed:.3f} with test/NDCG@10 = {score}")

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3430107	best: 0.3430107 (0)	total: 264ms	remaining: 44m 4s
200:	test: 0.4987203	best: 0.4987203 (200)	total: 38.5s	remaining: 31m 19s
400:	test: 0.5062413	best: 0.5062413 (400)	total: 1m 11s	remaining: 28m 38s
600:	test: 0.5100462	best: 0.5100462 (600)	total: 1m 42s	remaining: 26m 50s
800:	test: 0.5103095	best: 0.5111176 (780)	total: 2m 13s	remaining: 25m 35s
1000:	test: 0.5113879	best: 0.5115883 (920)	total: 2m 42s	remaining: 24m 23s
1200:	test: 0.5122962	best: 0.5125612 (1170)	total: 3m 12s	remaining: 23m 28s
1400:	test: 0.5134747	best: 0.5137121 (1370)	total: 3m 41s	remaining: 22m 39s
1600:	test: 0.5138865	best: 0.5140083 (1535)	total: 4m 11s	remaining: 21m 56s
1800:	test: 0.5144862	best: 0.5147300 (1760)	total: 4m 39s	remaining: 21m 14s
2000:	test: 0.5149401	best: 0.5152340 (1980)	total: 5m 9s	remaining: 20m 36s
2200:	test: 0.5149205	best: 0.5152340 (1980)	total: 5m 39s	remaining: 20m 2s
2400:	test: 0.5154017	best: 0.5154618 (2355)	total: 6m 8s	remaining: 19m 27s
2600:	t

Как видно, такой подход не дает прироста в качестве, но требует большое количество памяти (~29 гб) и сильно замедляет обучение. Остановимся на варианте все признаки + кластеризация.

***Итого: ничего не улучшило ситуацию.***

# Итоговый (грустный) результат

In [70]:
best_params

{'learning_rate': 0.10340626393986621,
 'depth': 7,
 'l2_leaf_reg': 17.272888261719643,
 'loss_function': 'YetiRank',
 'eval_metric': 'NDCG:top=10;type=Exp',
 'iterations': 10000,
 'bagging_temperature': 1.068369922077575,
 'use_best_model': True,
 'min_data_in_leaf': 68,
 'verbose': 200,
 'grow_policy': 'Depthwise',
 'task_type': 'GPU',
 'random_seed': 0}

In [71]:
class Model:
    def __init__(self, device='CPU'):
        params = {
            'learning_rate': 0.10340626393986621,
             'depth': 7,
             'l2_leaf_reg': 17.272888261719643,
             'loss_function': 'YetiRank',
             'eval_metric': 'NDCG:top=10;type=Exp',
             'iterations': 8000,
             'bagging_temperature': 1.068369922077575,
             'use_best_model': False,
             'min_data_in_leaf': 68,
             'verbose': 500,
             'grow_policy': 'Depthwise',
             'task_type': device,
             'random_seed': 0
        }
        self.model = catboost.CatBoost(params)

    def _to_catboost_dataset(self, df):
        y = df['label'].to_numpy()                       # Label: [0-4]
        q = df['qid'].to_numpy().astype('uint32')        # Query Id
        X = df.drop(columns=['label', 'qid']).to_numpy() # 136 features
        return (X, y, q)


    def fit(self, df_train, df_valid):
        X_train, y_train, q_train = self._to_catboost_dataset(df_train)
        X_valid, y_valid, q_valid = self._to_catboost_dataset(df_valid)
        X_train_valid = np.r_[X_train, X_valid]
        y_train_valid = np.r_[y_train, y_valid]
        q_train_valid = np.r_[q_train, q_valid]
        
        pool_train = catboost.Pool(data=X_train_valid, 
                                   label=y_train_valid, 
                                   group_id=q_train_valid
                                   )
        self.model.fit(pool_train)


    def predict(self, df_test):
        X_test, y_test, q_test = to_catboost_dataset(df_test)
        return self.model.predict(X_test)

Создадим и применим модель:

In [73]:
# Create model
model = Model('CPU')

# Fit
start = timer()
model.fit(df_train, df_valid)
elapsed = timer() - start
print(f"Model fit: elapsed = {elapsed:.3f}")

# Predict
y_hat_test = model.predict(df_test)
print(f"Predicted: y_hat_test.shape = {y_hat_test.shape}")

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	total: 43.9ms	remaining: 5m 50s
500:	total: 18.8s	remaining: 4m 40s
1000:	total: 36.7s	remaining: 4m 16s
1500:	total: 54.5s	remaining: 3m 55s
2000:	total: 1m 12s	remaining: 3m 36s
2500:	total: 1m 29s	remaining: 3m 17s
3000:	total: 1m 47s	remaining: 2m 59s
3500:	total: 2m 5s	remaining: 2m 41s
4000:	total: 2m 23s	remaining: 2m 23s
4500:	total: 2m 40s	remaining: 2m 5s
5000:	total: 2m 58s	remaining: 1m 47s
5500:	total: 3m 16s	remaining: 1m 29s
6000:	total: 3m 34s	remaining: 1m 11s
6500:	total: 3m 52s	remaining: 53.6s
7000:	total: 4m 10s	remaining: 35.7s
7500:	total: 4m 28s	remaining: 17.9s
7999:	total: 4m 46s	remaining: 0us
Model fit: elapsed = 291.476
Predicted: y_hat_test.shape = (241521,)


In [74]:
y_hat_test = model.predict(df_test)
print(f"Predicted: y_hat_test.shape = {y_hat_test.shape}")

Predicted: y_hat_test.shape = (241521,)


Теперь, имея предикты, можно посчитать метрики качества:

In [75]:
def compute_metrics(y_true, y_hat, q):
    # List of metrics to evaluate
    eval_metrics = ['NDCG:top=10;type=Exp']
    
    for eval_metric in eval_metrics:
        scores = utils.eval_metric(y_true, y_hat, eval_metric, group_id=q)
    
        # Print scores
        print(f"metric = {eval_metric} score = {scores[0]:.3f}")

# Get test targets and groups
y_test = df_test['label'].to_numpy()
q_test = df_test['qid'].to_numpy().astype('uint32')
    
# Compute metrics on test
compute_metrics(y_test, y_hat_test, q_test)

metric = NDCG:top=10;type=Exp score = 0.513


Ожидаем, что ваша модель покажет результаты выше бейзлайна!